Tutorial purview python sdk (#86524)
Azure/azure-content: articles/purview/tutorial-using-python-sdk.md

---
title: "How to use Azure Purview Python SDK"
description: This tutorial describes how to use the Azure Purview Python SDK to scan data and search the catalog.
author: raoufaliouat
ms.author: raoufaliouat
ms.service: purview
ms.subservice: purview-data-scanning
ms.topic: tutorial
ms.date: 01/10/2022

# Customer intent: I can use the scanning and catalog Python SDKs to perform CRUD operations on data sources and scans, trigger scans and also to search the catalog.
---

# Tutorial: Use the Purview Python SDK

The aim of this tutorial is to introduce you to the use of the Azure Purview Python SDK to programmatically do the most common operations you can find in Azure Purview Studio.

In this tutorial, you will learn how to:

> *   Grant the required rights to work programmatically with Purview

Reviewer: ```suggestion
*   Grant the required rights to work programmatically with Azure Purview
```

All instances of 'Purview' need to be updated to say 'Azure Purview' instead.

Reviewer: You'll need to make this sweep through the whole article.

*   Register a Blob Storage container as a data source in Purview
*   Define and run a scan
*   Search the catalog
*   Delete a data source

## Prerequisites

> For this tutorial, you will need :

Reviewer: For each of these, add a link the customer can follow if they DON'T have one of these prerequisites.

You can use the SQL Server tutorial as an example: https://docs.microsoft.com/en-us/azure/purview/tutorial-register-scan-on-premises-sql-server#prerequisites

*   Python 3.6+
*   An active Azure Subscription 
*   An Azure Active Directory tenant associated with you subscription
*   An Azure storage account
*   A Purview account

## Give Purview access to the Storage account

Before being able to scan the content of the Storage account, you need to give purview the right role. 

> Go to your Storage Account via the portal. Select Access Control (IAM). Click on the Add button and select “Add role assignment”.

Reviewer: ```suggestion
Go to your Storage Account via the portal. Select Access Control (IAM). Click on the Add button and select **Add role assignment**.
```

Bold these references to text in the UI, rather than using quotation marks. You'll need to update that through the article.

:::image type="content" source="media/tutorial-using-python-sdk/add-role-assignment-storage.png" alt-text="Screenshot that shows Access .Control menu of the Storage Account"::: 

In the next window, search for the “Storage blob Reader” role and select it:

:::image type="content" source="media/tutorial-using-python-sdk/storage-blob-reader-role.png" alt-text="Screenshot that shows the Storage Blob Reader role in the list of available roles"::: 

> Then go on the “Members” tab and click on “Select members”:

Reviewer: ```suggestion
Then go on the “Members” tab and select “Select members”:
```

Use 'Select' rather than 'click.'
You had already started doing this, so I'd suggest just doing a search for any rogue "click"s in your article.

:::image type="content" source="media/tutorial-using-python-sdk/select-members-blob-reader-role.png" alt-text="Screenshot that shows where to click to grant the role the Purview Managed System Identity"::: 

A new pane appears on the right. Search and select the name of your existing purview instance.
You can then click on "Review + Assign". Purview now has the required reading right to scan your Blob Storage. 

## Grant your application the access to your Purview account
1. First, you need a [service principal with a secret](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal#register-an-application-with-azure-ad-and-create-a-service-principal).

> Get the following information from your service principal and secret, they will be needed later:

Reviewer: Please add a screenshot showing where users can find this information.
Make sure to follow content guidelines to replace sensitive data, including keys, in your screenshots.

    *    Client ID (or Application ID)
    *    Tenant ID (or Directory ID)
    *    Client secret 
    

> 2. You now need to give the relevant Purview roles to your service principal. To do so, access the Purview instance. Select "Open Purview Studio" or open [Azure Purview's home page](https://web.purview.azure.com/) and choose the instance that you deployed.

Reviewer: ```suggestion
1. You now need to give the relevant Purview roles to your service principal. To do so, access the Purview instance. Select "Open Purview Studio" or open [Azure Purview's home page](https://web.purview.azure.com/) and choose the instance that you deployed.
```

Lists in markdown use 1. for every number in the list, and will automatically build out the correct numbers for you based on page structure. You'll need to make this change throughout the lists in your articles.

Author: I did not know that, thank you! :)

    Inside Purview Studio, go to the collections:
    :::image type="content" source="media/tutorial-using-python-sdk/purview-collections.png" alt-text="Screenshot that shows how to access Purview collections"::: 

>     Select the collection you want to work with, and go on the "Role assignments" tab. Add the service principal (using its name) in the following roles: Collection admins, Data source admins, Data curators, Data readers.

Reviewer: Please also provide a screenshot of at least one example of adding the service principal to one of these roles.

## Install the Python packages
1.	Open a new command prompt or terminal
2.	Install the Azure identity package for authentication:
    ```bash
    pip install azure-identity
    ```
3.	Install the Azure Purview Scanning Client package
    ```bash
    pip install azure-purview-scanning
    ```
4.	Install the Azure Purview Client package
    ```bash
    pip install azure-purview-catalog
    ```
5.	Install the Azure core package
    ```bash
    pip install azure-core
    ```

## Instantiate a Scanning and a Catalog client 
In this section, you learn how to instantiate:

> *    A scanning client useful to deal with scan-related tasks like registering data sources, creating and managing scan rules, triggering a scan, etc. 

Reviewer: ```suggestion
*    A scanning client useful to registering data sources, creating and managing scan rules, triggering a scan, etc. 
```

*    A catalog client useful to interact with the catalog through searching, browsing the discovered assets, identifying the sensitivity of your data, etc.

First you need to authenticate to your Azure Active Directory. In this tutorial, you will use the [client secret you created](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal#option-2-create-a-new-application-secret).

1.	Start with required import statements:
    ```python
    from azure.purview.scanning import PurviewScanningClient
    from azure.purview.catalog import PurviewCatalogClient
    from azure.identity import ClientSecretCredential 
    from azure.core.exceptions import HttpResponseError
    ```

> 2.	As said previously, you need to specify the following information in the code: 

Reviewer: ```suggestion
2.	Specify the following information in the code: 
```

    * Client ID (or Application ID)
    * Tenant ID (or Directory ID)
    * Client secret
    
    ```python
    client_id = "<your client id>" 
    client_secret = "<your client secret>"
    tenant_id = "<your tenant id>"
    ```

3.	You also need to specify the name of your purview account:
    ```python
    reference_name_purview = "<name of your purview account>"
    ```
4.	You can now instantiate the two clients:
    ```python
    def get_credentials():
        credentials = ClientSecretCredential(client_id=client_id, client_secret=client_secret, tenant_id=tenant_id)
        return credentials

    def get_purview_client():
        credentials = get_credentials()
        client = PurviewScanningClient(endpoint=f"https://{reference_name_purview}.scan.purview.azure.com", credential=credentials, logging_enable=True)  
        return client
    
    def get_catalog_client():
        credentials = get_credentials()
        client = PurviewCatalogClient(endpoint=f"https://{reference_name_purview}.purview.azure.com/", credential=credentials, logging_enable=True)
        return client
    ```
 

> You are now able to do several operations in a similar fashion.

Reviewer: Can you clarify this sentence? Are you going to show us how to do other things with a similar piece of code in the following sections? What kinds of things?

Author: I rephrased this sentence, is it clearer?

Reviewer: Unfortunately, I'm not seeing any change to this line.

                "resourceGroup": rg_name,
                "location": rg_location,
                "resourceName": storage_name,
                "resourceId": storage_id,
                "collection": {
                    "type": "CollectionReference",
                    "referenceName": reference_name_purview #here we use the root collection
                },
                "dataUseGovernance": "Disabled"
            }
        }    
    ```
    Note that in the above snippet, the name of the collection is the same as the Purview account because we work with the root collection that has the same name. Feel free to pass the name of an other collection you might have created.

3.	Register the DS
    ```python

    try:
        client = get_purview_client()
    except AzureError as e:
        print(e)

    try:
        response = client.data_sources.create_or_update(ds_name, body=body_input)
        print(response)
        print(f"Data source {ds_name} successfully created or updated")
    except HttpResponseError as e:
        print(e)
    ```

    When the operation succeeds, you can see an enriched body response from the client.

> All the operations are done in the same fashion. 

Reviewer: All what operations?

Author: I specified the operations in my latest commit, is it clearer?

    return credentials

def get_purview_client():
    credentials = get_credentials()
    client = PurviewScanningClient(endpoint=f"https://{reference_name_purview}.scan.purview.azure.com", credential=credentials, logging_enable=True)  
    return client

def get_catalog_client():
    credentials = get_credentials()
    client = PurviewCatalogClient(endpoint=f"https://{reference_name_purview}.purview.azure.com/", credential=credentials, logging_enable=True)
    return client

try:
    client = get_purview_client()
except AzureError as e:
    print(e)

try:
    response = client.data_sources.create_or_update(ds_name, body=body_input)
    print(response)
    print(f"Data source {ds_name} successfully created or updated")
except HttpResponseError as e:
    print(e)
```

## Scan the data source

Scanning a data source is done in three steps:
1.    (Optional): Define a Scan rule
2.    Create a scan definition
3.    Trigger a scan run

In this tutorial, you will use the default scan rules for Blob Storage containers. However, you can easily create custom scan rules programmatically with the Purview Scanning Client.

> Now let us scan the data source you registered above.

Reviewer: ```suggestion
Now let's scan the data source you registered above.
```
Microsoft likes contractions!
